subprojects/hashing/src/main/java/org/gradle/internal/hash/FileContentTypeDetectingInputStream.java

```diff
@@ -0,0 +1,117 @@
+/*
+ * Copyright 2021 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.hash;
+
+import java.io.IOException;
+import java.io.InputStream;
+
+/**
+ * Infers whether a file is a binary file or not by checking if there are any ASCII
+ * control characters in the file.  If so, then it is likely a binary file.
+ *
+ * Note that all operations should delegate to the underlying input stream, except
+ * for mark() and markSupported() which are unsupported.  The methods newly
+ * introduced in Java 9 (such as readAllBytes(), transferTo(), etc) are left
+ * unimplemented as this class currently must be Java 8 compatible.
+ */
```

Comment:
This text detection would probably work for ASCII and UTF-8 files, but would fail for UTF-16/UTF-32 encoded files with non-ASCII characters. For example the letter `ő` is written as `51 01` in UTF-16LE, and `01` would be detected as an ASCII control character, and hence the file would be treated as binary.

Reply:
See my comment at https://github.com/gradle/gradle/pull/17448#issuecomment-861727087.  I totally agree that this would be a problem for non-ascii encodings, but I feel like it is reasonable to just not support those encodings.